components/ai_chat/resources/page/components/markdown_renderer/index.tsx

```diff
@@ -0,0 +1,76 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import * as React from 'react'
+import Markdown from 'react-markdown'
+
+import styles from './style.module.scss'
+import CaretSVG from '../svg/caret'
+
+const CodeBlock = React.lazy(async () => ({
+  default: (await import('../code_block')).default.Block
+}))
+const CodeInline = React.lazy(async () => ({
+  default: (await import('../code_block')).default.Inline
+}))
+
+interface MarkdownRendererProps {
+  text: string
+  shouldShowTextCursor: boolean
+}
+
+export default function MarkdownRenderer(mainProps: MarkdownRendererProps) {
+  const [lastLine, setLastLine] = React.useState(1)
+
+  const plugin = React.useCallback(() => {
+    const transformer = (ast: any) => {
+      setLastLine(ast.position.end.line)
+    }
+
+    return transformer
+  }, [mainProps.text])
+
+  return (
+    <div className={styles.markdownContainer}>
+      <Markdown
+        rehypePlugins={[plugin]}
+        disallowedElements={['a', 'img']}
```

Comment:
Just curious, but could we allowlist elements instead of disallowing?

Reply:
yes, the markdown renderer has an allowlist API as well. however, I think we have more elements that we allow than disallow, such as paragraphs, strong, bold, table-related elements, etc. it seems much more manageable to specify what to disallow rather than what to allow.

Reply:
I agree that allowlisting is preferable as we won't be surprised by more unexpected elements especially if react-markdown adds new features in a future version. By scraping [syntax-tree/mdast-util-to-hast](https://github.com/syntax-tree/mdast-util-to-hast) which I think is the correct dependency to look in, the possible tags are:

```js
'a',
'blockquote',
'br',
'code',
'del',
'em',
'h1',
'h2',
'h3',
'h4',
'h5',
'h6',
'hr',
'img',
'li',
'li',
'ol',
'ol',
'p',
'pre',
'section',
'strong',
'sup',
'table',
'tbody',
'td',
'th',
'thead',
'tr',
'ul',
```

Potentially we also need to include `div` but not sure about that.

Reply:
ok, i have added an allowlist instead. excluded any table-related elements for now as we will support that on a follow-up PR.